Feature/auth service/creating service logic

Goodwin999 opened the pull request:
Вынес логирование пакета service в отдельный класс;
Написал логику JwtService ;
Написал логику AuthService;
Создал ключ для подписи токена;
Пришлось добавить репозиторий аналогичный другой ветке которая на PR;

Goodwin999 committed:
feature(auth):
created mail added it to application.yml;

Goodwin999 committed:
feature(auth):
created mail added it to application.yml;

Goodwin999 committed:
fix(auth):
removed email service;
added logging of the service package;

Goodwin999 committed:
fix(auth):
removed email service;
added logging of the service package;

Goodwin999 committed:
fix(auth):
fixed the ci crash

belousoveu commented on backend/auth-service/pom.xml:
```diff
@@ -91,6 +91,23 @@
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-api</artifactId>
+            <version>0.11.5</version>
```
версии в `<properties>`
и версия должна быть последняя

Goodwin999 replied:
Исправлено

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/entity/User.java:
```diff
@@ -30,6 +35,9 @@ public class User {
+    private boolean isActive = false;
+
     public enum Role implements GrantedAuthority {
         ADMIN, USER;
```
SERVICE добавь

Goodwin999 replied:
добавлено

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/entity/User.java:
```diff
@@ -30,6 +35,9 @@ public class User {
     @Enumerated(EnumType.STRING)
     private Role role;
 
+    @Column(name = "is_active", nullable = false)
```
убери пока. не до этого

Goodwin999 replied:
убрал

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/repository/UserRepository.java:
```diff
@@ -0,0 +1,21 @@
+    @Transactional
+    @Modifying
+    @Query("UPDATE User u SET u.isActive = true WHERE u.id = :userId")
+    void activateUser(Long userId);
```
тоже убери.
активацию делать не будем. не успеем

Goodwin999 replied:
убрал

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/AuthServiceImpl.java:
```diff
@@ -3,21 +3,73 @@
+    private final JwtService jwtService;
+    private final AuthenticationManager authenticationManager;
+
+    /**
```
что за мусор?

Goodwin999 replied:
убрал

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/loggerService/ServiceLogger.java:
```diff
@@ -0,0 +1,14 @@
+package com.cherkizon.auth.service.loggerService;
+
+import org.slf4j.Logger;
```
Зачем, когда это можно сделать одной аннотацией lombok???
Зачем "изобретать велосипед"?

Goodwin999 replied:
исправил на @UtilityClass

belousoveu replied:
:) Это неплохо, что ты узнал об еще одной аннотации Lombok
Но вообще-то я имел в виду другое. Зачем вообще нужен этот класс, когда ты всегда можешь поставить над классом, где тебе нужен логгер аннотацию `@Slf4j`? 
Лучше так и делать.

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/AuthServiceImpl.java:
```diff
@@ -3,21 +3,73 @@
+        AUTH.info("User registered: {}", user.getUsername());
     }
 
+    /**
```
Слава, зачем оставляешь мусор в коде?

Goodwin999 replied:
убрал

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/JwtServiceImpl.java:
```diff
@@ -1,11 +1,123 @@
+    private String secret;
+
+    @Value("${jwt.access-expiration}")
+    private long accessExpiration; // 24 часа
```
мусор не оставляй

Goodwin999 replied:
убрал

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/JwtServiceImpl.java:
```diff
@@ -1,11 +1,123 @@
+    @Override
+    public JwtResponse generateTokens(User user) {
+        Map<String, Object> claims = new HashMap<>();
+        claims.put("role", user.getRole().name());
```
Зачем писать три строки, чтобы отправить в качестве аргумента карту с одним значением?
Почему не `generateToken(Map.of("role",user.getRole().name()), user, accessExpiration)`,

Goodwin999 replied:
исправил

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/JwtServiceImpl.java:
```diff
@@ -1,11 +1,123 @@
+
+    private String generateToken(Map<String, Object> claims, UserDetails userDetails, long expiration) {
+        return Jwts.builder()
+                .setClaims(claims)
```
не надо использовать depicated методы никогда
Ну не ленись - посмотри в документации - какой метод новый вместо depricated

Goodwin999 replied:
исправил

Goodwin999 committed:
fix(auth):
added the port according to the instructions in application.yml

belousoveu commented on backend/auth-service/src/main/resources/application.yml:
```diff
@@ -12,6 +12,10 @@ spring:
+    jwt:
+      secret: ${JWT_SECRET}
+      access-expiration: 86400000 # 24 часа
+      refresh-expiration: 604800000 # 7 дней
```
Скажи мне - а зачем ты делаешь refresh на 7 дней, когда меняешь его одновременно с access токеном?

Goodwin999 replied:
переделал

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/JwtServiceImpl.java:
```diff
@@ -1,11 +1,123 @@
     public JwtResponse refreshToken(String refreshToken) {
-        // TODO: Реализовать логику
-        return null;
+        if (!isTokenValid(refreshToken)) {
```
недостаточная проверка для refresh
Во-первых почему ты его нигде не сохраняешь?
REfresh нужно хранить - когда присылают на продление пару - сравнивать c тем что хранится в базе
Как реализовано у тебя - я получу права Админа за минуту )

Goodwin999 replied:
исправил

belousoveu commented on backend/auth-service/src/main/resources/templates/activation-email.html:
```diff
@@ -0,0 +1,64 @@
+<!DOCTYPE html>
+<html lang="ru" xmlns:th="http://www.thymeleaf.org">
```
Это убирай. Если и успеем это сделать (в чем я пока сомневаюсь), то в любом случае делается это не так

Goodwin999 replied:
убрал

belousoveu reviewed (changes requested):
В целом нормально
1. Убирай пока все, что связано с активацией аккаунта.
2. Хранение и проверка refresh-токена. Хранить в крайнем случае можно в таблице user (что не очень хорошо). Лучше отдельную таблицу со связью OneToOne.
3. Мусор (ненужные комментарии) весь убери из кода.

Goodwin999 committed:
fix(auth):
added the port according to the instructions in application.yml

Goodwin999 commented:
> В целом нормально
> 
> 1. Убирай пока все, что связано с активацией аккаунта.
> 2. Хранение и проверка refresh-токена. Хранить в крайнем случае можно в таблице user (что не очень хорошо). Лучше отдельную таблицу со связью OneToOne.
> 3. Мусор (ненужные комментарии) весь убери из кода.

Планирую @ManyToOne потому что один пользователь может иметь несколько токенов (например, вошёл с телефона и с браузера),
@OneToOne будет ограничивать нас одним токеном на пользователя 
Остальное выношу в отдельную сущность Token + репозиторий для него

Goodwin999 committed:
fix(auth):
corrected all comments, added entity token, and script;

the application runs on port 8081;

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/JwtServiceImpl.java:
```diff
@@ -1,11 +1,135 @@
+                .revoked(false)
+                .build();
+
+        tokenRepository.save(newToken);
```
А почему в этот момент старый токен мы не удаляем? Зачем продолжает хранить в БД. С какой целью?
Через какое-то время у нас база будет забита просроченными токенами (которые кстати при этом по состоянию полей `expired` и `revoked` будут считаться активными - это уже нарушение логической структуры)
Либо аргументируй зачем мы храним эти просроченные токены либо удаляй

belousoveu replied:
И хранить их лучше все-таки в закодированном виде (как пароли) через passwordEncoder

Goodwin999 replied:
исправил

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/JwtServiceImpl.java:
```diff
@@ -1,11 +1,135 @@
-        return null;
+        if (!isTokenValid(refreshToken)) {
+            JWT.error("Invalid refresh token attempt");
+            throw new InvalidTokenException("Invalid refresh token");
```
Смотри - если токен не действительный, то у тебя  в этом методе `isTokenValid(refreshToken)` уже выбрасывается исключение и поскольку ты проверку не оборачиваешь в try-catch, то у тебя блок if - вообще никогда не выполняется же. Зачем он тут?

Goodwin999 replied:
исправил это, и комментарий выше по @Slf4j

belousoveu replied:
А где исправил? ничего же не исправлено

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/JwtServiceImpl.java:
```diff
@@ -1,11 +1,135 @@
+        Token savedToken = tokenRepository.findByRefreshToken(refreshToken)
+                .orElseThrow(() -> new InvalidTokenException("Refresh token not found or already rotated"));
+
+        if (savedToken.isExpired() || savedToken.isRevoked()) {
```
у тебя поля `expired` и `revoked` нигде в программе не меняются - как создались вместе с токеном, так и создались.
Соответственно вопросы 
- зачем тогда эти поля нужны?
- какого результата проверки ты ждешь от этого if?

Goodwin999 replied:
удалил поля, удалил проверку

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/JwtServiceImpl.java:
```diff
@@ -1,11 +1,135 @@
+        Instant now = Instant.now();
+        return Jwts.builder()
+                .claims(claims)
+                .subject(userDetails.getUsername())
```
А почему subject - username? 
нужен userId

Goodwin999 replied:
испраил на userId

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/entity/Token.java:
```diff
@@ -0,0 +1,30 @@
+    @Column(nullable = false)
+    private boolean expired;
+
+    @Column(nullable = false)
```
Это поле не нужно если оно нигде не используется

Goodwin999 replied:
убрал

belousoveu replied:
Тоже вот оно поле, как было - так и есть
![image](https://github.com/user-attachments/assets/7e64be91-be93-4fa7-98f7-037085cfe04b)

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/entity/Token.java:
```diff
@@ -0,0 +1,30 @@
+    private User user;
+
+    @Column(nullable = false)
+    private boolean expired;
```
Это поле в таком виде тоже бессмысленно
Лучше сделать поле типа Instant и хранить в нем время истечения токена

Goodwin999 replied:
убрал

belousoveu replied:
Ты пишешь, что убрал - но я вижу, что оно осталось. Это как?

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/entity/User.java:
```diff
@@ -14,6 +15,7 @@
+@Builder
 @EqualsAndHashCode(onlyExplicitlyIncluded = true)
 @ToString(exclude = "password")
 public class User implements UserDetails {
```
В этот класс тогда нужно тоже поле добавить с отношением oneToMany и настроить каскадное удаление, т.е. чтобы при удалении пользователя сразу удалялись все его refresh токены
Т.е. физически у тебя нового поля в таблице users не должно добавляться, но внешний ключ прописать и настроить - надо

Goodwin999 replied:
исправил, файл миграции сейчас поправлю

Goodwin999 replied:
p.s
точнее новый создам(старый нельзя трогать)

belousoveu reviewed (changes requested):
Смотри, по рефреш-токену
OneToMany - так OneToMany - ок
Но делать из базы "кладбище" просроченных токенов - считаю нецелесообразно
Алгоритм должен быть такой:
1. При Регистрации выдаем пару токенов. Рефреш при этом сохраняем в таблицу
2. При обновлении порядок действий такой
  - ищем в таблице - если нет пользователь идет лесом
  - если есть, но просрочен - тоже идет лесом. пусть заново логинится. При этом токен из таблицы удаляем
  - Получаем usera из токена, сравниваем (на всякий случай - совпадает ли он с юзером на котором числится этот токен в таблице) - если нет - опять таки - в лес. токен из базы удаляем
  - формируем пару новых токенов. Рефреш записываем в таблицу. Старый токен из таблицы удаляем
  - возвращаем пару новых токенов
 **Запись нового токена и удаление старого - должны выполняться в одной трансакции**
Остальные замечания - по тексту

belousoveu commented:
Если ты исправил - коммить - пока я никаких изменений не вижу

Goodwin999 committed:
fix(auth):
Bug fixes;
added token expiration time field;

Goodwin999 committed:
Merge branch 'dev' into feature/auth-service/creating_service_logic

Goodwin999 committed:
fix(auth):
Bug fixes;

Goodwin999 committed:
chore(auth):
added migration file to create token lifetime in DB;

Goodwin999 committed:
chore(auth):
added migration file to create token lifetime in DB;

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/JwtServiceImpl.java:
```diff
@@ -1,11 +1,148 @@
+            throw new InvalidTokenException("Invalid refresh token");
+        }
+
+        Token savedToken = tokenRepository.findAll().stream()
```
Что за дикое обращение с БД? зачем выбирать все записи, чтобы потом в стриме найти одну?
Почему сразу не взять из таблицы нужную запись?
Ну уж в крайнем случае выбрать по Юзеру токены

Goodwin999 replied:
добавил List<Token> userTokens = tokenRepository.findAllByUserId(userId);

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/JwtServiceImpl.java:
```diff
@@ -1,11 +1,148 @@
+        Long userId = extractUserId(rawRefreshToken);
+        User tokenUser = savedToken.getUser();
+
+        if (!tokenUser.getId().equals(userId)) {
```
id - примитивы же. Зачем equals?

Goodwin999 replied:
возвращается Long - объект обертка для id

Goodwin999 replied:
private Long id; — это обёртка, не примитив

Goodwin999 committed:
chore(auth):
added migration file to create token lifetime in DB;

Goodwin999 committed:
chore(auth):
applied migration to token table

Goodwin999 committed:
chore(auth):
applied migration to token table

belousoveu reviewed (changes requested):
А ты мои последние замечания не видел?
обрати внимание

Goodwin999 committed:
fix(auth):
fix bag

Goodwin999 committed:
feat(auth): add JWT support

Goodwin999 committed:
feat(auth): add JWT support

Goodwin999 committed:
fix(auth): add JWT support

Goodwin999 committed:
fix(auth): add JWT support
